From 947f8017767cf0c9aea57ec18413cfd3ed6bad30 Mon Sep 17 00:00:00 2001
From: Richard Zhu <hongxing.zhu@nxp.com>
Date: Mon, 27 Nov 2017 13:06:15 +0800
Subject: [PATCH 2907/5242] MLK-16982 PCI: imx: fix the failure of the msi
 verification

commit  ab40ff3c60d4eb82eb037fd97ca0e73612c033e1 from
https://source.codeaurora.org/external/imx/linux-imx.git

Failed to verify the MSI in the EP RC system.
Root cause: the MSI address is not fetched corretly.
The second port of iMX8MQ EVK board should be used
as EP port, not the first one.

Signed-off-by: Richard Zhu <hongxing.zhu@nxp.com>
Reviewed-by: Frank Li <frank.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/pci/controller/dwc/pci-imx6-ep-driver.c |   12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/pci/controller/dwc/pci-imx6-ep-driver.c b/drivers/pci/controller/dwc/pci-imx6-ep-driver.c
index dd300a9..4ea0782 100644
--- a/drivers/pci/controller/dwc/pci-imx6-ep-driver.c
+++ b/drivers/pci/controller/dwc/pci-imx6-ep-driver.c
@@ -51,10 +51,10 @@ static int imx_pcie_ep_probe(struct pci_dev *pdev,
 		const struct pci_device_id *id)
 {
 	int ret = 0, index = 0, found = 0;
-	unsigned int msi_addr = 0, cpu_base;
+	unsigned int hard_wired = 0, msi_addr = 0, cpu_base;
 	struct resource cfg_res;
 	const char *name = NULL;
-	struct device_node *np;
+	struct device_node *np = NULL;
 	struct device *dev = &pdev->dev;
 	struct imx_pcie_ep_priv *priv;
 
@@ -90,7 +90,13 @@ static int imx_pcie_ep_probe(struct pci_dev *pdev,
 		goto err_pci_unmap_mmio;
 	}
 
-	np = of_find_compatible_node(NULL, NULL, "snps,dw-pcie");
+	/* Use the first none-hard-wired port as ep */
+	while ((np = of_find_node_by_type(np, "pci"))) {
+		if (of_property_read_u32(np, "hard-wired", &hard_wired)) {
+			hard_wired = 0;
+			break;
+		}
+	}
 	if (of_property_read_u32(np, "cpu-base-addr", &cpu_base))
 		cpu_base = 0;
 
-- 
1.7.9.5

